src/query/storages/common/index/tests/it/bloom_pruner.rs

```diff
@@ -477,17 +532,44 @@ fn eval_index_expr(
         expr
     };
 
-    let fields = bloom_columns.values().cloned().collect::<Vec<_>>();
-    let (_, scalars) = BloomIndex::filter_index_field(&expr, &fields).unwrap();
+    let bloom_fields = bloom_columns.values().cloned().collect::<Vec<_>>();
+    let ngram_fields = ngram_args
+        .iter()
+        .map(|arg| arg.field().clone())
+        .collect::<Vec<_>>();
+    let result = BloomIndex::filter_index_field(&expr, bloom_fields, ngram_fields).unwrap();
 
     let mut scalar_map = HashMap::<Scalar, u64>::new();
-    for (scalar, ty) in scalars.into_iter() {
+    for (_, scalar, ty) in result.bloom_scalars.into_iter() {
         scalar_map.entry(scalar).or_insert_with_key(|scalar| {
             BloomIndex::calculate_scalar_digest(&func_ctx, scalar, &ty).unwrap()
         });
     }
 
-    let mut builder = BloomIndexBuilder::create(func_ctx.clone(), bloom_columns.clone());
+    for (field, (_, scalar)) in result
+        .ngram_fields
+        .iter()
+        .zip(result.ngram_scalars.into_iter())
+    {
+        let Some(ngram_arg) = ngram_args.iter().find(|arg| arg.field() == field) else {
+            continue;
+        };
+        for words in BloomIndex::calculate_ngram_nullable_column(
+            Value::Scalar(scalar),
+            ngram_arg.gram_size(),
+            ngram_arg.bitmap_size(),
+            |w, _| w.to_string(),
+        ) {
+            for word in words {
+                let hash = BloomIndex::ngram_hash(&word, ngram_arg.bitmap_size());
+
+                scalar_map.entry(Scalar::String(word)).or_insert(hash);
```

Comment:
don't put ngram scalar string in `scalar_map`, the word may be conflict with scalar string in bloom scalar, which will result in wrong result.
add a new map to store ngram scalar string map and use the string as the key, not the splitted word, so that we don't need to caculate hash values in each blocks.